lib/Sema/TypeCheckAvailability.cpp

```diff
@@ -1967,36 +1967,56 @@ void TypeChecker::checkConcurrencyAvailability(SourceRange ReferenceRange,
   }
 }
 
+/// Returns the diagnostic to emit for the potentially unavailable decl and sets
+/// \p IsError accordingly.
+static Diagnostic getPotentialUnavailabilityDiagnostic(
+    const ValueDecl *D, const DeclContext *ReferenceDC,
+    const UnavailabilityReason &Reason, bool WarnBeforeDeploymentTarget,
+    bool &IsError) {
+  ASTContext &Context = ReferenceDC->getASTContext();
+  auto Platform = prettyPlatformString(targetPlatform(Context.LangOpts));
+  auto Version = Reason.getRequiredOSVersionRange().getLowerEndpoint();
+
+  const bool RequiredVersionIsDeploymentTargetOrLower =
+      Version <= AvailabilityContext::forDeploymentTarget(Context)
+                     .getOSVersion()
+                     .getLowerEndpoint();
+
+  if (RequiredVersionIsDeploymentTargetOrLower) {
+    IsError = !WarnBeforeDeploymentTarget;
+
+    return Diagnostic(
+        IsError ? diag::availability_decl_only_version_newer_for_clients
+                : diag::availability_decl_only_version_newer_for_clients_warn,
+        D->getName(), ReferenceDC->getParentModule(), Platform, Version);
```

Comment:
What is the goal of printing the reference module name here?

I'm wondering if printing it could be more confusing than helpful in the case where the unavailable decl comes from a different module. You could see a `View (as in SwiftUI.View) is only available to clients of MyLib in macOS 12.0 or newer`, when `View` is actually available more widely to the clients importing `SwiftUI` directly. Maybe we should talk about the interface of the reference module: `View is only available in the API/module interface of MyLib in macOS 12.0`.

Reply:
Or should we reference `D`'s module instead?

Reply:
I think that using `D`'s module name would not convey what I'm trying to convey. When this diagnostic is emitted, `D` would always be available to code in the module being built. The issue is that it might not be available to the clients of the module being built. The clients of `D`'s module, however, include the module being built. So if the diagnostic were reworded to refer to `D`'s module then it is not adding any semantic meaning over the previous diagnostic wording.

Reply:
I've update the diagnostic text to this:
```
'Foo' is only available in macOS X.Y or newer; clients of 'Module' may have a lower deployment target
```